### ISSUE TITLE
updateSubscription missing fields

### ISSUE THREAD

**Issue by stepcut** [dmjio/stripe#10]
I am trying to port my code from my own stripe library to this one. The API allows you to update many aspects of the subscription:

https://stripe.com/docs/api/curl#update_subscription

But the current implementation of the updateSubscription function only allows you to modify the coupon and metadata. Unfortunately, my code needs to modify which plan they are subscribed to. Not sure why only some of the fields are supported and no way is provided to update the other fields.

The same is true of many API calls. The author of this library has arbitrarily decided which parts of the stripe API are available rather than expose the full API.

**Comment by alexwestholm** [comment]
Thanks for the module, @dmjio! Any movement on this? Running into similar issues. Off the cuff, @stepcut's approach seems to strike a good balance between internal complexity and ease of external use.

**Comment by dmjio** [comment]
@alexwestholm, @stepcut changes will represent the newest API. They just need to be merged.

**Comment by jonschoning** [comment]
When arguments get too complex, one solution is to create a data type just for the arguments. Then you can create a default constructor for that type, and fill in only what you want to replace.  That seems more self-documenting to me than sprinkling constraints around in instance declarations

**Comment by dmjio** [comment]
@jonschoning, That's true, but it's not that simple. That was my initial approach, but let's flesh that out. There is a `metadata` field for at least 11 different areas (Invoices, Subscription, Customer), etc. This means both the `Invoice` JSON return object and `InvoiceParameters` records would have metadata fields aptly named `invoiceMetaData` and `invoiceParameterMetaData`. Since haskell records don't allow you to name the same field twice, I now need to introduce at least 22 fields that have the same type with slightly different naming conventions, that makes for one ugly API. The answer is obviously lens, but now I introduce a cognitive overhead to noobies and incur a massive dependency overhead. If you install this package from scratch you'll see that the Types.hs file takes ~10 seconds to compile (There's a ton of types). If I were to go down this path  that time would only increase. 

Another downside to the default record approach is that I'm assuming I know what defaults are appropriate for your application, I'm bound to upset most. By choosing the various functions that I did I attempted to cover most commonly used scenarios and if all else fails, expose a function that takes every parameter and let the user customize it.

I think @stepcut 's approach is the correct one. Have all API functions take the required parameters and allow the user to specify the level of granularity they need via a typeclass instance. It fits the problem well. Instead of multiple `metadata` fields, I now have a metadata typeclass, of which, 11+ records are instances of, and the API is far cleaner. 

The problem here is that I have a full time job and this merge creates 3 packages, 1 of which I am not the maintainer of yet, and requires a lot of synchronization between people, and upgrades to the new Stripe 2015 API.
